src/client.ts

```diff
@@ -165,95 +149,151 @@ export class SubscriptionClient {
   }
 
   public unsubscribe(id: number) {
-    delete this.subscriptions[id];
-    delete this.waitingSubscriptions[id];
-    let message = { id: id, type: SUBSCRIPTION_END};
-    this.sendMessage(message);
+    if (this.requests[id]) {
+      delete this.requests[id];
+    }
+    this.sendMessage(id, MessageTypes.GQL_END, undefined);
   }
 
   public unsubscribeAll() {
-    Object.keys(this.subscriptions).forEach( subId => {
-      this.unsubscribe(parseInt(subId));
+    Object.keys(this.requests).forEach( subId => {
+      this.unsubscribe(parseInt(subId, 10));
     });
   }
 
-  // send message, or queue it if connection is not open
-  private sendMessage(message: Object) {
-    switch (this.client.readyState) {
+  private executeRequest(options: RequestOptions, handler: (error: Error[], result?: any) => void): number {
```

Comment:
Actually, I think I'm fine with `operationOptions`. But another point here. If all operations can be long running, we need a way to cancel them, so `executeRequest` could for example return a function, which when called will stop the operation, just like it works with observables.